bpo-16500: Don't use string constants for os.register_at_fork() behavior

gpshead opened the pull request:
Add `os` module constants and updates all uses and documentation instead of string literals.
This allows for cleaner code with less chance of a typo in a string value showing up at runtime.
Linters and similar tools can flag incorrect constant names before run time.

gpshead committed:
Use os module constants for os.register_at_fork.

Adds the constants and updates all uses and documentation.

pitrou commented:
I'm not sure I like this. This is longer to type, and doesn't seem to bring anything over human-readable string constants (compare with, say, `SEET_SET` and friends which are opaque integer values). If you make a typo, since `os.register_at_fork` will typically be called at module import, that typo will probably show up immediately.

@1st1 ?

gpshead commented:
I wouldn't expect `register_at_fork()` to be called at import time in most non-stdlib use cases. An application is more likely to call it later as it starts running things that need special treatment (especially for crazy developers who refuse to learn that mixing fork and threads is a bad idea).

My main complaint is the use of magic string values.

Alternatives would be
 * To make three functions instead of having a when parameter.
 * Or to mirror pthread_atfork()'s API and accept three callables (keyword args only at that point). I kind of like that idea. A single register_at_fork() function, with three optional named parameters.

pitrou commented:
Well, I would rather have constants than your two other suggestions :-)

gpshead commented:
```python
os.register_at_fork(before=prepare_for_launch,
                    after_parent=launch_cleanup,
                    after_child=reinitialize_stuff)
```

Looks very readable to me.

pitrou commented:
Why not. It's much better than having three separate functions, at least.

gpshead committed:
Switch to keyword only arguments instead of consts

This switches from using str constants or os module constants
on a warn parameter to using named keyword only arguments
specifying the intent of the callable being passed in.  It avoids
the need for string literals or new os module constants and is
generally more readable.  This is more similar to how the libc
pthread_atfork() API works.

gpshead committed:
Undo the unnecessary blank line os.py change.

gpshead committed:
Fix a docstring typo in posixmodule.

serhiy-storchaka commented on Doc/library/os.rst:
```diff
@@ -3280,16 +3280,20 @@ written in Python, such as a mail server's external command delivery program.
+   * *after_in_parent* is a function called from the parent process
+     after forking a child process.
+   * *after_in_child* is a function called from the child process
+     if control is expected to return to the Python interpreter.  A
```
Is "if control is expected to return to the Python interpreter" related to other arguments? `_posixsubprocess.fork_exec()` doesn't call any of handlers if preexec_fn is None.

And the same question about the documentation of `PyOS_AfterFork_Child()` vs other functions.

serhiy-storchaka commented on Lib/test/test_posix.py:
```diff
@@ -189,19 +189,29 @@ def test_waitid(self):
+        with self.assertRaises(TypeError, msg="Positional args not allowed"):
+            os.register_at_fork(lambda: None)
+        with self.assertRaises(TypeError, msg="Args must be callable"):
+            os.register_at_fork(before=2, after_in_child="three",
```
It is not clear what argument triggered an error. Use three different invocations with three `assertRaises`.

serhiy-storchaka commented on Lib/test/test_posix.py:
```diff
@@ -189,19 +189,29 @@ def test_waitid(self):
+            os.register_at_fork(before=2, after_in_child="three",
+                                after_in_parent=b"Five")
+        with self.assertRaises(TypeError, msg="Args must not be None"):
+            os.register_at_fork(before=None, after_in_child=None,
```
The same as above.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5312,49 +5312,50 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
-'before' callbacks are called in reverse order before forking.
-'child' callbacks are called in order after forking, in the child process.
-'parent' callbacks are called in order after forking, in the parent process.
+``before`` callbacks are called in reverse order before forking.
```
Double backticks don't look nice in text output.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5312,49 +5312,50 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
-    if (!PyCallable_Check(func)) {
-        PyErr_Format(PyExc_TypeError,
-                     "expected callable object, got %R", Py_TYPE(func));
+    if (!(PyCallable_Check(before) || PyCallable_Check(after_in_child) ||
```
This means that `os.register_at_fork(before=lambda:None, after_in_child=42, after_in_parent='spam')` is successful.

I would change this condition to just `(!(before || after_in_child || after_in_parent))` or `(!before && !after_in_child && !after_in_parent)` and add separate `PyCallable_Check()` checks for every non-NULL argument.

Since this case was not caught by tests, needed a new test for it.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5312,49 +5312,50 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
-        PyErr_Format(PyExc_ValueError, "unexpected value for `when`: '%s'",
-                     when);
+    if (before &&
+        register_at_forker(&interp->before_forkers, before)) {
```
The check for `before != NULL` can be moved into `register_at_forker()` together with a `PyCallable_Check()` check. Unless you want to produce more specific error messages.

serhiy-storchaka commented on Doc/c-api/sys.rst:
```diff
@@ -49,10 +49,12 @@ Operating System Utilities
+   or any similar function that clones the current process, if there is
+   any chance the process will call back into the Python interpreter.
+   Only available on systems where :c:func:`fork` is defined as determined
+   by the ``HAVE_FORK`` C preprocessor define.
```
`HAVE_FORK` is internal CPython define. It is not a part of any API.

1st1 commented:
Having linters and IDEs helping you with the `register_at_fork` signature is a good idea, although I still like the original API proposed by @pitrou.

+1 to go with keyword arguments, instead of constants.

```python
os.register_at_fork(before=prepare_for_launch,
                    after_parent=launch_cleanup,
                    after_child=reinitialize_stuff)
```

gpshead committed:
Don't mention the HAVE_FORK cpp #define.

gpshead committed:
Don't use backticks in the docstring.

gpshead committed:
bpo-16500, don't call forkers if we never forked().

We don't want to call PyOS_BeforeFork() or PyOS_AfterFork_Parent()
if we never manage to make it to the fork() call it self, erroring
out instead.

gpshead committed:
bpo-16500: Clarify when all three calls are made.

The statement about re-entering the interpreter applies to all three
calls, not just after_in_child.

gpshead committed:
Improve the test cases to cover the required logic.

gpshead committed:
fix register_at_fork parameter validation.

gpshead commented:
Thanks @serhiy-storchaka I believe I've addressed all of those suggestions.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5309,52 +5311,68 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
+check_null_or_callable(PyObject *obj, const char* obj_name)
+{
+    if (obj && !PyCallable_Check(obj)) {
+        PyErr_Format(PyExc_TypeError, "'%s' must be callable, not %R",
```
The error message looks like:
```
'before' must be callable, not <class 'str'>
```
It is more common to include the type name rather than repr.
```
'before' must be callable, not str
```

gpshead replied:
%s and Py_TYPE(obj)->tp_name solves that. thanks.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5309,52 +5311,68 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
 #ifdef HAVE_FORK
+
+/* Helper function to validate arguments.
+   Returns 0 on success.  1 on failure with a TypeError raised.
```
It is more common to return either
* 1 on success, 0 on failure, or
* 0 on success, -1 on failure.

gpshead replied:
it doesn't really matter.  0 on success and non-zero on failure is the real idiom.  1 and -1 are both non-zero and pass a truth test.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5309,52 +5311,68 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
-Register a callable object to be called when forking.
+Register a callable to be called when forking a new process via os.fork().
 
 'before' callbacks are called in reverse order before forking.
```
This partially repeats descriptions of arguments ('before' is called before forking, etc).

gpshead replied:
I removed some of the redundancy in the docstring.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5309,52 +5311,68 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
-        'before', 'child' or 'parent'
+    *
+    before: object=NULL
+        Zero arg callable to be called in the parent before the fork() syscall.
```
"in the parent" perhaps is redundant.

And "Zero arg" may be redundant in the docstring. This detail can be explained in the module documentation. Shortened "arg" is a jargon word.

serhiy-storchaka commented on Modules/posixmodule.c:
```diff
@@ -5309,52 +5311,68 @@ os_spawnve_impl(PyObject *module, int mode, path_t *path, PyObject *argv,
+        Zero arg callable to be called in the parent before os.fork() returns.
 
-Register a callable object to be called when forking.
+Register a callable to be called when forking a new process via os.fork().
```
Callable**s**.

gpshead committed:
update threading to use the new API.

serhiy-storchaka commented on Doc/library/os.rst:
```diff
@@ -3280,16 +3280,21 @@ written in Python, such as a mail server's external command delivery program.
-   is forked.  *when* is a string specifying at which point the function is
-   called and can take the following values:
+   Register callables to be executed when a new child process is forked
+   using :func:`os.fork`.  The parameters are optional and keyword-only.
```
`os.fork()` is not the only forking function. There are also `os.fork1()`, `os.forkpty()` and `_posixsubprocess.fork_exec()`. And third-party code can call `fork()`.

gpshead replied:
It is impossible and unmaintainable to list them all, I've generalized the wording.

gpshead committed:
Documentation refactoring based on code review.

gpshead committed:
Use the type name rather than repr in the error.

serhiy-storchaka commented:
I don't know why, but the diff now shows already merged changes, including unrelated to this issue. This makes reviewing hard.

gpshead commented:
I don't know why git did that. I followed the https://cpython-devguide.readthedocs.io/gitbootcamp.html#syncing-with-upstream instructions to pull in recent changes so the tests would pass and I could update Lib/threading.py to use this API. :(

gpshead commented:
Okay, I figured out how to do a proper rebasing and the diffs for this PR look sane again without including merged changes from others.

(for reference, the git instructions I followed were based on https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request)

pitrou commented on Modules/clinic/posixmodule.c.h:
```diff
@@ -1828,40 +1828,44 @@ os_spawnve(PyObject *module, PyObject **args, Py_ssize_t nargs, PyObject *kwname
 "--\n"
 "\n"
-"Register a callable object to be called when forking.\n"
+"Registers callables to be called when forking a new process.\n"
```
As a style rule, CPython uses infinitives in documentation, not present tense.

pitrou commented:
btw, if you don't want to mess around with git rebases, `git pull origin master` should be sufficient to merge the latest changes from master into your current branch (replace "origin" with whatever the upstream CPython is named in your working clone).

gpshead committed:
Use the infinitive in the docstring.
